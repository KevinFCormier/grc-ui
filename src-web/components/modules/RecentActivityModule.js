/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
'use strict'

import React from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'
import resources from '../../../lib/shared/resources'
import msgs from '../../../nls/platform.properties'
import _ from 'lodash'
//import { Link } from 'react-router-dom'

resources(() => {
  require('../../../scss/module-recent-activity.scss')
})

export default class RecentActivityModule extends React.Component {

  render() {
    const { locale } = this.context
    const title = msgs.get('overview.recent.activity.title', locale)
    const { handleDrillDownClick } = this.props
    const data = this.getModuleData()
    return (
      <div className='module-recent-activity'>
        <div className='card-container-container'>
          <div className='card-container'>
            <div className='card-content'>
              <div className='card-inner-content'>
                <div className='card-title'>
                  {title}
                </div>
                {data.violations.map(({count, violationType}, idx) => {
                  const onClick = () =>{
                    handleDrillDownClick(violationType)
                  }
                  const onKeyPress = (e) =>{
                    if ( e.key === 'Enter') {
                      onClick()
                    }
                  }
                  const classes = classNames({
                    'card-count-type': true,
                    hasBorder: idx===1,
                  })
                  return (
                    <div key={violationType} className={classes} role={'button'}
                      tabIndex='0' onClick={onClick} onKeyPress={onKeyPress}>
                      <div className='card-count'>
                        {count}
                      </div>
                      <div className='card-type'>
                        <div>
                          {violationType.toUpperCase()}
                        </div>
                        <div>
                          {msgs.get('overview.recent.activity.violation.type', locale).toUpperCase()}
                        </div>
                      </div>
                    </div>
                  )
                })}
                {data.findings.map(({count, findingType}) => {
                  const onClick = () =>{
                    handleDrillDownClick(findingType)
                  }
                  const onKeyPress = (e) =>{
                    if ( e.key === 'Enter') {
                      onClick()
                    }
                  }
                  return (
                    <div key={findingType} className='card-count-type' role={'button'}
                      tabIndex='0' onClick={onClick} onKeyPress={onKeyPress}>
                      <div className='card-count'>
                        {count}
                      </div>
                      <div className='card-type'>
                        <div>
                          {msgs.get('overview.recent.activity.finding.severity', [findingType], locale).toUpperCase()}
                        </div>
                        <div>
                          {msgs.get('overview.recent.activity.finding.type', locale).toUpperCase()}
                        </div>
                      </div>
                    </div>
                  )
                })}
              </div>
            </div>
          </div>
        </div>
      </div>
    )
  }

  getModuleData = () => {
    const { locale } = this.context
    const { policies } = this.props
    const policySet = new Set()
    const clusterSet = new Set()
    policies.map(policy=>{
      const statuses = _.get(policy, 'raw.status.status', {})
      Object.keys(statuses).forEach(key=>{
        const compliant = statuses[key].compliant
        if (!compliant || compliant.toLowerCase()==='noncompliant') {
          clusterSet.add(key)
          policySet.add(_.get(policy, 'metadata.name', 'unknown'))
        }
      })
    })
    return {
      violations: [
        {count: policySet.size, violationType: msgs.get('overview.top.violations.policy', locale)},
        {count: clusterSet.size, violationType: msgs.get('overview.top.violations.cluster', locale)},
      ],
      findings: [
        {count: 3, findingType: 'High'},
        {count: 23, findingType: 'Medium'},
        {count: 28, findingType: 'Low'},
      ]
    }
  }
}

RecentActivityModule.propTypes = {
  handleDrillDownClick: PropTypes.func,
  policies: PropTypes.array,
}
