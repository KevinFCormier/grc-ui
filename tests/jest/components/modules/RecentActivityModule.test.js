/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018, 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
'use strict'

import React from 'react'
import { BrowserRouter } from 'react-router-dom'
import renderer from 'react-test-renderer'
import { policiesTestingDataSet1, findingsTestingDataSet1 } from './ModuleTestingData'
import RecentActivityModule from '../../../../src-web/components/modules/RecentActivityModule'

describe('RecentActivityModule view', () => {
  const handleDrillDownClick = jest.fn()
  const filteredPolicies = policiesTestingDataSet1
  const filteredFindings = findingsTestingDataSet1
  const viewState = {}
  const updateViewState = jest.fn()
  it('renders as expected', () => {
    const component = renderer.create(<BrowserRouter><RecentActivityModule
      policies={filteredPolicies}
      findings={filteredFindings}
      viewState={viewState}
      updateViewState={updateViewState}
      handleDrillDownClick={handleDrillDownClick} /></BrowserRouter>)
    expect(component).toMatchSnapshot()
  })
})
