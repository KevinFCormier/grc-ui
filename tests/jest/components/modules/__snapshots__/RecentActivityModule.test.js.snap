// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecentActivityModule view renders as expected 1`] = `
<div
  className="module-recent-activity"
>
  <div
    className="card-container-container"
  >
    <div
      className="card-title"
    >
      Recent activity
    </div>
    <div
      className="card-container"
    >
      <div
        className="card-content"
      >
        <div
          className="card-inner-content"
        >
          <div
            className="card-count-type"
            onClick={[Function]}
            onKeyPress={[Function]}
            role="button"
            tabIndex="0"
          >
            <div
              className="card-count alert"
            >
              3
            </div>
            <div
              className="card-type"
            >
              <div>
                POLICY VIOLATIONS
              </div>
            </div>
          </div>
          <div
            className="card-count-type hasBorder"
            onClick={[Function]}
            onKeyPress={[Function]}
            role="button"
            tabIndex="0"
          >
            <div
              className="card-count alert"
            >
              2
            </div>
            <div
              className="card-type"
            >
              <div>
                CLUSTER VIOLATIONS
              </div>
            </div>
          </div>
          <div
            className="card-count-type"
            onClick={[Function]}
            onKeyPress={[Function]}
            role="button"
            tabIndex="0"
          >
            <div
              className="card-count-container"
            >
              <svg
                className="card-count-pie"
              >
                <g>
                  <path
                    d="M9.18485099360515e-16,-15A15,15,0,1,1,-9.18485099360515e-16,15A15,15,0,1,1,9.18485099360515e-16,-15M-1.6532731788489267e-15,-9A9,9,0,1,0,1.6532731788489267e-15,9A9,9,0,1,0,-1.6532731788489267e-15,-9Z"
                    fill="#E62325"
                  />
                  <path
                    d="M-2.3880612583373385e-15,-13L-2.0206672185931327e-15,-11Z"
                    fill="#DFE3E6"
                  />
                </g>
              </svg>
              <div
                className="card-count alert"
              >
                3
              </div>
            </div>
            <div
              className="card-type"
            >
              <div>
                HIGH SEVERITY
              </div>
            </div>
          </div>
          <div
            className="card-count-type"
            onClick={[Function]}
            onKeyPress={[Function]}
            role="button"
            tabIndex="0"
          >
            <div
              className="card-count-container"
            >
              <svg
                className="card-count-pie"
              >
                <g>
                  <path
                    d="M9.18485099360515e-16,-15L5.51091059616309e-16,-9Z"
                    fill="#FEC233"
                  />
                  <path
                    d="M7.960204194457795e-16,-13A13,13,0,1,1,-7.960204194457795e-16,13A13,13,0,1,1,7.960204194457795e-16,-13M-2.0206672185931327e-15,-11A11,11,0,1,0,2.0206672185931327e-15,11A11,11,0,1,0,-2.0206672185931327e-15,-11Z"
                    fill="#DFE3E6"
                  />
                </g>
              </svg>
              <div
                className="card-count"
              >
                0
              </div>
            </div>
            <div
              className="card-type"
            >
              <div>
                MEDIUM SEVERITY
              </div>
            </div>
          </div>
          <div
            className="card-count-type"
            onClick={[Function]}
            onKeyPress={[Function]}
            role="button"
            tabIndex="0"
          >
            <div
              className="card-count-container"
            >
              <svg
                className="card-count-pie"
              >
                <g>
                  <path
                    d="M9.18485099360515e-16,-15L5.51091059616309e-16,-9Z"
                    fill="#A0A0A0"
                  />
                  <path
                    d="M7.960204194457795e-16,-13A13,13,0,1,1,-7.960204194457795e-16,13A13,13,0,1,1,7.960204194457795e-16,-13M-2.0206672185931327e-15,-11A11,11,0,1,0,2.0206672185931327e-15,11A11,11,0,1,0,-2.0206672185931327e-15,-11Z"
                    fill="#DFE3E6"
                  />
                </g>
              </svg>
              <div
                className="card-count"
              >
                0
              </div>
            </div>
            <div
              className="card-type"
            >
              <div>
                LOW SEVERITY
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="violation-container"
    >
      <div
        className="module-top-violations"
      >
        <div
          className="header-container"
        >
          <div
            className="header-title"
          >
            Top violations
          </div>
          <div
            className="header-tab"
          >
            <nav
              aria-label="Top violations undefined"
              className="bx--tabs"
              role="navigation"
              selected={1}
            >
              <div
                aria-label="listbox"
                className="bx--tabs-trigger"
                onClick={[Function]}
                onKeyPress={[Function]}
                role="listbox"
                tabIndex={0}
              >
                <a
                  className="bx--tabs-trigger-text"
                  href="#"
                  onClick={[Function]}
                  tabIndex={-1}
                >
                  Clusters
                </a>
                <svg
                  alt="show menu options"
                  aria-label="show menu options"
                  className={undefined}
                  fill={undefined}
                  fillRule="evenodd"
                  height="5"
                  name={undefined}
                  role="img"
                  style={undefined}
                  viewBox="0 0 10 5"
                  width="10"
                >
                  <title>
                    show menu options
                  </title>
                  <path
                    d="M0 0l5 4.998L10 0z"
                  />
                </svg>
              </div>
              <ul
                className="bx--tabs__nav bx--tabs__nav--hidden"
                role="tablist"
              >
                <li
                  className="bx--tabs__nav-item"
                  onClick={[Function]}
                  onKeyDown={[Function]}
                  role="presentation"
                  selected={false}
                  tabIndex={-1}
                >
                  <a
                    aria-selected={false}
                    className="bx--tabs__nav-link"
                    href="#"
                    role="tab"
                    tabIndex={-1}
                  >
                    Policies
                  </a>
                </li>
                <li
                  className="bx--tabs__nav-item bx--tabs__nav-item--selected"
                  onClick={[Function]}
                  onKeyDown={[Function]}
                  role="presentation"
                  selected={true}
                  tabIndex={-1}
                >
                  <a
                    aria-selected={true}
                    className="bx--tabs__nav-link"
                    href="#"
                    role="tab"
                    tabIndex={0}
                  >
                    Clusters
                  </a>
                </li>
              </ul>
            </nav>
            <div
              aria-hidden={true}
              className={undefined}
              hidden={true}
              selected={false}
            />
            <div
              aria-hidden={false}
              className={undefined}
              hidden={false}
              selected={true}
            />
          </div>
        </div>
        <div>
          <div
            className="violation-card-container"
          >
            <div>
              <div
                className="violation-card-container"
              >
                <div
                  className="violation-card-content"
                >
                  <div
                    className="card-inner-content"
                  >
                    <div
                      className="card-violations"
                      onClick={[Function]}
                      onKeyPress={[Function]}
                      role="button"
                      tabIndex="0"
                    >
                      <div
                        className="card-violation-count alert"
                      >
                        3
                      </div>
                    </div>
                    <div
                      className="card-name-description"
                    >
                      <div
                        className="card-name"
                      >
                        cluster1
                      </div>
                      <div
                        className="card-description"
                      >
                        <a
                          className="card-each-description-link"
                          href="undefined/policies/policy/policy-comp/compliance-all"
                          onClick={[Function]}
                        >
                          compliance-all
                        </a>
                        , 
                        <a
                          className="card-each-description-link"
                          href="undefined/policies/policy/policy-comp/policy-ma"
                          onClick={[Function]}
                        >
                          policy-ma
                        </a>
                        , 
                        <a
                          className="card-each-description-link"
                          href="undefined/policies/policy/policy-comp/policy-pod"
                          onClick={[Function]}
                        >
                          policy-pod
                        </a>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div>
              <div
                className="violation-card-container"
              >
                <div
                  className="violation-card-content"
                >
                  <div
                    className="card-inner-content"
                  >
                    <div
                      className="card-violations"
                      onClick={[Function]}
                      onKeyPress={[Function]}
                      role="button"
                      tabIndex="0"
                    >
                      <div
                        className="card-violation-count alert"
                      >
                        2
                      </div>
                    </div>
                    <div
                      className="card-name-description"
                    >
                      <div
                        className="card-name"
                      >
                        clusterhub
                      </div>
                      <div
                        className="card-description"
                      >
                        <a
                          className="card-each-description-link"
                          href="undefined/policies/policy/policy-comp/compliance-all"
                          onClick={[Function]}
                        >
                          compliance-all
                        </a>
                        , 
                        <a
                          className="card-each-description-link"
                          href="undefined/policies/policy/policy-comp/policy-pod"
                          onClick={[Function]}
                        >
                          policy-pod
                        </a>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="module-top-violations"
      >
        <div
          className="header-container"
        >
          <div
            className="header-title"
          >
            Top security findings
          </div>
          <div
            className="header-tab"
          >
            <nav
              aria-label="Top security findings undefined"
              className="bx--tabs"
              role="navigation"
              selected={0}
            >
              <div
                aria-label="listbox"
                className="bx--tabs-trigger"
                onClick={[Function]}
                onKeyPress={[Function]}
                role="listbox"
                tabIndex={0}
              >
                <a
                  className="bx--tabs-trigger-text"
                  href="#"
                  onClick={[Function]}
                  tabIndex={-1}
                >
                  Security findings
                </a>
                <svg
                  alt="show menu options"
                  aria-label="show menu options"
                  className={undefined}
                  fill={undefined}
                  fillRule="evenodd"
                  height="5"
                  name={undefined}
                  role="img"
                  style={undefined}
                  viewBox="0 0 10 5"
                  width="10"
                >
                  <title>
                    show menu options
                  </title>
                  <path
                    d="M0 0l5 4.998L10 0z"
                  />
                </svg>
              </div>
              <ul
                className="bx--tabs__nav bx--tabs__nav--hidden"
                role="tablist"
              >
                <li
                  className="bx--tabs__nav-item bx--tabs__nav-item--selected"
                  onClick={[Function]}
                  onKeyDown={[Function]}
                  role="presentation"
                  selected={true}
                  tabIndex={-1}
                >
                  <a
                    aria-selected={true}
                    className="bx--tabs__nav-link"
                    href="#"
                    role="tab"
                    tabIndex={0}
                  >
                    Security findings
                  </a>
                </li>
                <li
                  className="bx--tabs__nav-item"
                  onClick={[Function]}
                  onKeyDown={[Function]}
                  role="presentation"
                  selected={false}
                  tabIndex={-1}
                >
                  <a
                    aria-selected={false}
                    className="bx--tabs__nav-link"
                    href="#"
                    role="tab"
                    tabIndex={-1}
                  >
                    Clusters
                  </a>
                </li>
              </ul>
            </nav>
            <div
              aria-hidden={false}
              className={undefined}
              hidden={false}
              selected={true}
            />
            <div
              aria-hidden={true}
              className={undefined}
              hidden={true}
              selected={false}
            />
          </div>
        </div>
        <div>
          <div
            className="violation-card-container"
          >
            <div>
              <div
                className="violation-card-container"
              >
                <div
                  className="violation-card-content"
                >
                  <div
                    className="card-inner-content"
                  >
                    <div
                      className="card-violations"
                      onClick={[Function]}
                      onKeyPress={[Function]}
                      role="button"
                      tabIndex="0"
                    >
                      <div
                        className="card-violation-count alert"
                      >
                        3
                      </div>
                    </div>
                    <div
                      className="card-name-description"
                    >
                      <div
                        className="card-name"
                      >
                        <a
                          className="card-name-link"
                          href="undefined/policies/all/kube-system/Policy%20that%20is%20not%20compliant"
                          onClick={[Function]}
                        >
                          Policy that is not compliant
                        </a>
                      </div>
                      <div
                        className="card-description"
                      >
                        <a
                          className="card-each-description-link"
                          href="undefined/policies/policy/mycluster_name/undefined"
                          onClick={[Function]}
                        >
                          mycluster_name
                        </a>
                        , 
                        <a
                          className="card-each-description-link"
                          href="undefined/policies/policy/cluster1/undefined"
                          onClick={[Function]}
                        >
                          cluster1
                        </a>
                        , 
                        <a
                          className="card-each-description-link"
                          href="undefined/policies/policy/clusterhub/undefined"
                          onClick={[Function]}
                        >
                          clusterhub
                        </a>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
